modules/srm-client/man/man1/srmcp.1
.TH SRMCP "1" "May 2005" "SRM Fermi Client ....  1_6" "User Commands"
.SH NAME
srmcp \- manual page for srmcp v1_6
.SH SYNOPSIS
.B srmcp
[\fIOPTION\fR]... \fISOURCE DEST\fR
.br
.B srmcp
[\fIOPTION\fR]... \fISOURCE\fR... \fIDIRECTORY\fR
.br
.B srmcp
[\fIOPTION\fR]... \fI-copyjobfile <file>\fR...
.SH DESCRIPTION
Copy SOURCE to DEST, or Multiple SOURCE(s) to DIRECTORY.

Either source(s) or destination or both should be (an) srm url
default options will be read from configuration file 
but can be overridden by the command line options
the command line options are one or more of the following:
.TP
\fB\-\-version\fR, \fB\-version\fR 
enables printing version number
.TP
\fB\-\-debug\fR, \fB\-debug\fR
enables debug output, default is false 
.TP
\fB\-srmcphome\fR=<path to srmcp product dir>
.TP
\fB\-gsissl\fR=<true or false> 
true uses gsi https, default is false
.TP
\fB\-wsdl_url\fR=<full url to web service wsdl> 
this options overrides \fB\-webservice_path\fR and \fB\-webservice_protocol\fR options
.TP
\fB\-webservice_path\fR=<webservice_path> 
to specify web service path component of web service URL (for example "srm/managerv1.wsdl")
.TP
\fB\-webservice_protocol\fR=<webservice_protocol> 
to specify the webservice protocol ( for example "http","https" or "httpg")
.TP
\fB\-use_proxy\fR=<true or false> 
true if srmcp should use grid proxy false if it should use certificate and key directly, defult value is true
.TP

\fB\-x509_user_proxy\fR=<path to user grid proxy>
.TP
\fB\-x509_user_cert\fR=<path to user grid certificate>
.TP
\fB\-x509_user_key\fR=<path to user private key>
.TP
\fB\-x509_user_trusted_certificates\fR=<path to the directory with cerificates of trusted Certificate Authorities (CAs)>
.TP
\fB\-conf\fR=<path to the configuration file> default value is config.xml
.TP
\fB\-save_conf\fR=<path to the file in which the new configuration will be saved>
no transfer will be performed if this option is specified
.TP
.TP
\fB\-retry_timeout\fR=<number of miliseconds to sleep after a failure before the client tries to retry>
.TP
\fB\-retry_num\fR=<number of retries before client gives up>
.TP
\fB\-connect_to_wsdl\fR=<true or false, false by default> 
srm client now connects directly to the service without reading the wsdl first but for the compatibility with the old implementation, especially if srm urls available point to the location of the wsdl, we make the old way of connecting to the wsdl location first available
.TP
\fB\-delegate\fR=<true or false, true by default> 
enables or disables the delegation of the user credenital to the server
.TP
\fB\-full_delegation\fR=<true or false, true by default> 
if delegation is enabled, it specifies the type of delegation if this option is set to true the delegation is full, otherwise limited
.TP
\fB\-h\fR or \fB\-help\fR for this help 
.TP
\fB\-urlcopy\fR=<urlcopy path> 
to specify the path to  universal url_copy script see $SRM_PATH/bin/url-copy.sh for example
.TP
\fB\-buffer_size\fR=<integer> to set the buffer size to a value 
      different then default(2048)
.TP
\fB\-tcp_buffer_size\fR=<integer> 
to set the tcp buffer size to a value if option is not specified or set to 0, then the default tcp buffer size is used
.TP
\fB\-streams_num\fR=<integer> to set the number of streams used by gridftp 
.TP
\fB\-protocols\fR=protocol1[,protocol2[...]] 
the comma separated list of supported TURL protocols
.TP
\fB\-pushmode\fR=<true or false>  
true to use the push mode in case of srm Mass Storage Systems (MSS) to MSS copy, false to use the pull mode, the default mode is pull mode (false)
.TP
\fB\-use_urlcopy_script\fR=<true or false> 
use java native copiers of use urcopy script
.TP
\fB\-copyjobfile\fR=<file> 
where <file> is the path to the text file containing  the list of sources and destination each line has a format : <sorce-url> <destination-url>
.TP
\fB\-report\fR=<report_file> 
where <report_file> is the path to the report file if specified, it will contain the resutls of the execution srmcp the each line in the file will have the following format:

       <src url> <dst url> <return code> [<error>]

.PP
      the following return codes are supported:
      0 - success
      1 - general error
      2 - file exists, can not overwrite
      3 - user permission error

Example of srm put:
      srmcp file:////bin/sh srm://myhost.mydomain.edu:8443//dir1/dir2/sh-copy

Example of srm get:
      srmcp srm://myhost.mydomain.edu:8443//dir1/dir2/sh-copy file:///localdir/sh
Example of srm copy (srm to srm):
      srmcp srm://myhost.mydomain.edu:8443//dir1/dir2/sh-copy srm://anotherhost.org:8443/newdir/sh-copy

Example of srm copy (gsiftp to srm):
      srmcp gsiftp://ftphost.org//path/file srm://myhost.mydomain.edu:8443//dir1/dir2/file
.SH AUTHOR
Written by Timur Perelmutov 
.SH "REPORTING BUGS"
Report bugs to <timur@fnal.gov>
.SH COPYRIGHT
Storage Resource Manager (SRM) CP Client version 1.16
Copyright (c) 2002-2005 Fermi National Accelerator Laborarory




ChangeLog
2010-10-07 14:27  pmillar

	* src/Makefile.am: dcap: fix building on Solaris
	  
	  A previous commit (committed as r14667) broke compilation on
	  Solaris machines when using the
	  Solaris compiler. This patch fixes that problem.
	  
	  Target: trunk
	  Patch: http://rb.dcache.org/r/2487/
	  Acked-by: Owen

2010-10-07 13:33  pmillar

	* src/Makefile.am, src/dccp.c, src/print_size.c, src/print_size.h,
	  src/sigfig.c, src/sigfig.h: dcap: add -H option to dccp that
	  provides feedback during transfer
	  
	  Whilst in Karlsruhe for GridKa school, I was talking to someone
	  about their experience of dCache
	  being very slow when copying files into and from dCache via dccp.
	  The user was unable to answer
	  simple questions that might indicate what is the problem (where
	  you waiting for a mover? was the
	  checksum taking a long time? was the transfer slow on-average, or
	  did it "stutter"?). I released
	  that this information is currently difficult (or impossible) to
	  obtain with dccp. Moreover, without
	  any feedback during a transfer, one must have blind faith that
	  the transfer is proceeding.
	  
	  This patch addresses this issue. It provides some feedback to the
	  user during a transfer. The
	  feedback uses the (old) method of writing output without the
	  line-feed character ('\n') but using the
	  carriage-return character ('\r') instead. Repeated use of such
	  output gives the user the impression
	  of a single line of text is being updated with new information.
	  Some care is needed to append spaces
	  to the end of the new version of the line if it has shrunk. This
	  is to hide character(s) from the
	  previous line.
	  
	  The transfer feedback line contains a progress bar followed by
	  zero or more items. The items are:
	  percentage of file transferred, number of bytes transferred and
	  the average transfer rate. The code
	  determines the width (number of columns) of the terminal
	  dynamically and scales the output
	  accordingly. If the terminal has sufficient width then all items
	  are displayed and the bar is scaled
	  to take (almost) all of the line's remaining capacity. If the
	  width is too small to allow this then
	  items are progressively switched off to provide additional space,
	  allowing the progress bar to grow
	  in size. This continues until a minimum width of the progress bar
	  is achieved or all items are
	  switched off (whichever happens first).
	  
	  The progress bar is a sequence of characters the first being the
	  open square-bracket and the last
	  being the close square-bracket. Those in between are either a
	  dash ('-') or a hash ('#') and form
	  the part that indicates the transfer's progress.
	  
	  If the size of the file being transferred is known in advance,
	  the progress bar is a "percentage
	  progress bar". This shows information as a sequence of zero or
	  more hash symbols followed by zero or
	  more dashes. The total number of symbols is fixed, for given a
	  progress bar size. The number of
	  hash- and dash- symbols is chosen to be proportional to the
	  amount of data written and still to write
	  respectively. The result is an ASCII-art version of the familiar
	  GUI progress-bar. It is also how
	  the rpm command shows progress.
	  
	  The following is an example line when the file size is known and
	  the terminal is sufficiently wide to
	  allow all items:
	  
	  [#######---------------------------------------------------------------]
	  11% 45.0 MiB (4.09 MiB/s)
	  
	  If the size of the transfer isn't known in advance (e.g., data is
	  taken from stdin) then it is
	  impossible to show a "percentage activity bar". Instead, the
	  progress bar is an "activity progress
	  bar". This indicates bytes transferred as a shorter bar (made up
	  of some five hashes). This shorter
	  bar is initially left-most within the progress bar and "moves to
	  the right" as data is written. When
	  the shorter bar reaches the end of the activity bar, it wraps.
	  (This similar to how Firefox shows
	  progress under similar circumstances.)
	  
	  The following is an example when the transfer size is unknown and
	  the terminal is sufficiently wide
	  to allow all items:
	  
	  [-----#####----------------------------------------------------------------]
	  8.00 MiB (4.00 MiB/s)
	  
	  In writing this support, I adding a function for formatting file
	  sizes using SI binary prefixes (kiB,
	  MiB, etc), taking care of aesthetics. The patch also updates the
	  final line of output (which gives a
	  summary of the transfer) to also use the new function.
	  
	  The following is an example of current dccp's final output line:
	  
	  417488294 bytes in 78 seconds (5226.97 KB/sec)
	  
	  This is an example of final output line after patch:
	  
	  417488294 bytes (398 MiB) in 76 seconds (5.24 MiB/s)
	  
	  Current limitations and future work:
	  o no user feedback when the transfer stalls,
	  o could add an estimate of the time remaining until the transfer
	  finishes.
	  
	  Target: trunk
	  Patch: http://rb.dcache.org/r/2400/
	  Acked-by: Gerd
	  Acked-by: Owen
	  

2010-09-23 08:08  pmillar

	* src/dcap_error.c: dcap: fix type casting of type-punning
	  
	  In dcap_error.c, the support for thread-local variable,
	  t_getspecific macro,
	  which is a simple wrapper around pthread_getspecific(), is used
	  in dcap_error to
	  obtain a ptr to the thread-local errno. The dcap_error casts this
	  pointer to a
	  (void *), rather than (int *).
	  
	  Assigning to a type-punning, like this, is not allowed by C89/C90
	  standard but
	  historically, compilers turn a blind eye to them.
	  
	  It's legal to assign a (void *) point to any other kind of
	  pointer without
	  casting. The pthread_getspecific() function returns a (void *)
	  pointer, so the
	  explicit casts in dcap_error are actually unnecessary.
	  
	  The proposed patch fixes the "dereferencing type-punned pointer
	  will break
	  strict-aliasing rules" warnings.
	  

modules/srmclient/srmTools/srmcp.cpp
//
// srmcp client for Storage Resource Manager, v1.0
// put/get file to/from SRM server
//
//
// Usage: 
//  srmcp [options] <file> <file>
//  srmcp --help
//    to see list of currently available options
//
// Alex Kulyavtsev, FNAL
//

#include "config.h"
#include "srmWSH.h"
#include "stdio.h"
#include <sys/time.h>

#include "stdlib.h"
#include "unistd.h"
#include <sys/wait.h>

#include "SrmSoapWsClient.hpp"
#include "SrmWsClientConfig.hpp"
#include "url.hpp"

#include <vector>
#include <iterator>
#include <string>
#include <strstream>

// some include files in gsi.h conflict with #include <string>, so it must go after it
//   and "SrmSoapWsClient.hpp":
#include "gsi.h"

//---------------
using namespace srm;

  // Keep track of the file CVS version in the binary file
  static char cvsid[] = "$Id: srmcp.cpp,v 1.8 2003-11-03 16:03:26 cvs Exp $";


class srmURL : public url {
public:

  enum {
    SrmSchemeUnknown=0, // default on construction
    SrmSchemeFILE,
    SrmSchemeFTP,
    SrmSchemeKFTP,
    SrmSchemeGSIFTP,
    SrmSchemeHTTP,
    SrmSchemeHTTPS,
    SrmSchemeSRM,
    SrmSchemeMAX
  } srmScheme_t;

  srmURL( const string& u )
    : url( u ) {};
  int getScheme();
};

int srmURL::getScheme()
{
  int schNum = SrmSchemeUnknown;
  string sch = this->scheme();

  if ( sch == "file" )        schNum = SrmSchemeFILE;
  else if ( sch == "ftp" )    schNum = SrmSchemeFTP;
  else if ( sch == "kftp" )   schNum = SrmSchemeKFTP;
  else if ( sch == "gsiftp" ) schNum = SrmSchemeGSIFTP;
  else if ( sch == "http" )   schNum = SrmSchemeHTTP;
  else if ( sch == "https" )  schNum = SrmSchemeHTTPS;
  else if ( sch == "srm" )    schNum = SrmSchemeSRM;
    
  return schNum;
}

int
main (int argc, char *argv[] )
{
  //------------------
  // srmcp client data:

  int delay   = 1; // set initial delay
  //---------------

  int      i;
  int      ret;
  int      all_done = 0;
  int      errors = 0;
  string   rqState;
  string   fileState;

  //  srmInt   requestId;
  RequestStatus *prs, *prsDone = 0;
  string   funcName = "some-call-to-srm()";

  int retCode = -1;
  //----------------
  int   ind;
  class SrmWsClientConfig *cfg = new SrmWsClientConfig();

  //--------------------
  ind = cfg->parseCLOptions0( argc, argv );

  ret = cfg->readConfig();
  if ( ret ) {
    cerr << "Can not read configuration, exiting" << endl;
    exit(1);
  }

  ind = cfg->parseCLOptions( argc, argv );

  if ( cfg->srmDebug ){
    cout << "Configuration: " << endl;
    cfg->printConfig();  // Print updated configuration:
  }

  //--------------------

  if ( cfg->urlCopyScript == "" ){
    cerr << "srmcp(): urlCopyScript is not defined, exiting" << endl;
    exit(1);
  }
  const char *executable = cfg->urlCopyScript.c_str();

  //----------------
  int narg = argc -ind;
  if ( narg < 2 || narg > 2 ) {
    if ( narg < 2  ) 
      cerr << "Too few arguments. Must be src and dst at least. Exiting. " << endl << endl;
    if ( narg > 2  ) 
      cerr << "Too many arguments. Exiting. " << endl << endl;
    cfg->usage( "srmcp" );
    exit(1);
  }
  char ** argf;
  argf = argv + ind;

  class srmURL *u1 = new srmURL( argf[0]);
  class srmURL *u2 = new srmURL( argf[1]);
  class srmURL *srmURL  = NULL;
  class srmURL *fileURL = NULL;

  // Binary mask,          // srm is 
  const int doLOCAL= 0;    //        ...not involved, both are local files
  const int doGET  = 1;    //        ...Source
  const int doPUT  = 2;    //        ...Destination
  const int doCOPY = 3;    //        ... both, (doGET | doPUT) - bitwise OR

  int toDo = doLOCAL;
  int schN;

  schN = u1->getScheme();
  if ( schN == srmURL::SrmSchemeSRM )
    toDo |= doGET;
  else if ( schN != srmURL::SrmSchemeFILE ) {
    cout << "Protocol=" << schN << " is not served, exiting" << endl;
    cout << "Must be 'srm:' or 'file:'" << endl;
    cout << "URL=" << u1->getURL() << endl;
    exit(1);
  }

  schN = u2->getScheme();
  if ( schN == srmURL::SrmSchemeSRM )
    toDo |= doPUT;
  else if ( schN != srmURL::SrmSchemeFILE ) {
    cout << "Protocol=" << schN << " is not served, exiting" << endl;
    cout << "Must be 'srm:' or 'file:'" << endl;
    cout << "URL=" << u2->getURL() << endl;
    exit(1);
  }

  switch ( toDo ) {
  case doGET:
    srmURL  = u1;
    fileURL = u2;
    funcName= "get()";
    break;
  case doPUT:
    fileURL = u1;
    srmURL  = u2;
    funcName= "put()";
    break;

  case doLOCAL:
    cout << "Will not copy file: to file:, exiting" << endl;
    exit(1);
  case doCOPY:
    cout << "Will not copy srm: to srm:, exiting" << endl;
    exit(1);
  default:
    cout << "srmcp(), internal error - can't decide to get() or put() ... exiting" << endl;
    exit(1);
  }

  //----------------
  string srmProtocol = cfg->webServiceProtocol;
  string srmHName = srmURL->host();
  string srmPort  = srmURL->port();
  string srmPath  = cfg->webServicePath;
  class SrmHostInfo srmHostInfo = SrmHostInfo( srmProtocol, srmHName, srmPort, srmPath );

  string srmHost = srmHostInfo.getInfo();

  if ( cfg->srmDebug )
    cout << "SRM Host Info =" << srmHost << endl;

  //---------------------------------
  // Init the gsoap runtime environment

  class SrmSoapWsClient *cs = new SrmSoapWsClient();

  if ( cfg->useGsiSsl ) {
    if ( cfg->srmDebug )
      cout << "Use GsiSSL" << endl;
      
    if ( cs->setupGSI() ) {
      cerr << "Can not register GSI plugin, exiting" << endl;
      exit (1);
    }

    if ( cs->setSrmURL( srmHost ) ) {
      cerr << "Can not set SRM URL, exiting" << endl;
      exit (1);
    }
  }
  //==============================================

  class VArrayOfstring protocols( cfg->srmProtocols );

  vector<string> v_srcs(1);
  vector<string> v_dsts(1);
  vector<long> v_size(1);      // or   vector<LONG64> v_size(2);
  vector<bool> v_perm(1);

  v_srcs[0] = argf[0]; // fix me,  v_srcs.push_back( argf[j] );
  v_dsts[0] = argf[1]; // fix me

  v_size[0] = 0; // File size (for space reservation) - not impl. - fix me
  v_perm[0] = 0; // Make file permanent - not impl.

  //---
  class VArrayOfstring  srcs(  v_srcs );
  class VArrayOfstring  dsts(  v_dsts );
  class VArrayOflong    sizes( v_size );
  class VArrayOfboolean permanent( v_perm );
  //---

  if ( cfg->srmDebug )
    cout << " call "<< funcName << " - wait for connection " << endl;

  switch ( toDo ) {
  case doGET:
    ret = cs->get( &srcs, &protocols, &prs );
    break;
  case doPUT:
    ret = cs->put( &srcs, &dsts, &sizes, &permanent, &protocols, &prs );
    break;
  default:
    cout << "srmcp(), internal error - toDo switch ... exiting" << endl;
    exit(1);
  }

  if ( cfg->srmDebug )
    cout << " Ret code for " << funcName << " = " << ret << endl;

  if( ret ) {
    cerr << " " << funcName << ": got error #" << ret << ", exiting" << endl;
    exit(1);
  }
  if( prs == NULL ) {
    cerr << " srmcp(): internal error - " << funcName << " return OK,"
      " but pointer to ret.stat. is NULL, exiting" << endl;
    exit(1);
  }


  //====================
  // Request sent. 
  // Do not exit until file status set to Done

  if ( cfg->srmDebug ) {
    cout << "requestId = " << prs->requestId << endl;
    cout << "type      = " << prs->type      << endl;
    cout << "state     = " << prs->state     << endl;
    if(  string(prs->state) == "Failed" )
          cout << "error     = " << prs->errorMessage << endl;
    for(i=0 ; i < prs->fileStatuses->__size; i++ ) {
      cout << "  FILE # " << i << " ID fileId = " << prs->fileStatuses->__ptr[i].fileId << endl;
      cout << "  FILE # " << i << " ID size   = " << prs->fileStatuses->__ptr[i].size << endl;
      cout << "  FILE # " << i << " ID SURL   = " << prs->fileStatuses->__ptr[i].SURL << endl;
    }
    cout.flush();
  }

  // prs already checked for nonzero 
  if (   string(prs->state) != "Failed" ) {
//     requestId = prs->requestId;
    all_done = 0;

    // Wait for SRM to get TURLs
    while( ! all_done ) {

      delay = prs->retryDeltaTime;
      if ( cfg->srmDebug ) 
	cout << " sleep " << delay << " sec ..." << endl << flush;
      sleep( delay );

      ret = cs->getRequestStatus( prs->requestId, &prs );
      if( ret ) {
	cerr << " getRequestStatus(): got error " << ret << ", exiting" << endl;
	errors++;
	goto Exit;
      }
      if( prs == NULL ) {
	cerr << " getRequestStatus(): internal error - getRequestStatus() return OK,"
	  " but pointer to ret.stat.==NULL, exiting" << endl;
	errors++;
	goto Exit;
      }

      // Scan request is done
      all_done = 1;
      if ( string(prs->state) != "Failed" ) {
	for( i=0; i < prs->fileStatuses->__size; i++ ) {
	  fileState = prs->fileStatuses->__ptr[i].state;
	  if(  fileState == "Pending" ) 
	    all_done = 0;
	}
      }

      //=============
      if ( cfg->srmDebug ) {
	cout << " Request ID= " << prs->requestId << endl;
	cout << " type      = " << prs->type      << endl;
	cout << " state     = " << prs->state     << endl;
	if(  string(prs->state) == "Failed" )
          cout << "error     = " << prs->errorMessage << endl;
	cout << " submit Time = "        << prs->submitTime << endl;
    
	for( i=0 ; i < prs->fileStatuses->__size; i++ ) {
	  cout << "  FILE # " << i << " ID fileId = " << prs->fileStatuses->__ptr[i].fileId << endl;
	  cout << "  FILE # " << i << " ID size   = " << prs->fileStatuses->__ptr[i].size << endl;
	  cout << "  FILE # " << i << " ID SURL   = " << prs->fileStatuses->__ptr[i].SURL << endl;
	  cout << "  FILE # " << i << " ID TURL   = " << prs->fileStatuses->__ptr[i].TURL << endl;
	  cout << "  FILE # " << i << " ID state  = [" << prs->fileStatuses->__ptr[i].state 
	       << "] " << endl;
	}

	if ( all_done )
	  cout << "All done =" << all_done << endl << flush;
      }
      //=============
    }

    // Scan if request done
    if ( string(prs->state) == "Failed" ) {
      cout << "Request Failed" << endl;      
      cout << "error     = " << prs->errorMessage << endl;
      errors++;
    }else{
      prsDone = 0;
      for( i=0; i < prs->fileStatuses->__size; i++ ) {
	fileState = prs->fileStatuses->__ptr[i].state;
	if( fileState != "Ready" ){
	  errors++;
	  if( fileState != "Done" ){  // That is for "Failed" or "Pending"
	    ret = cs->setFileStatus( prs->requestId, prs->fileStatuses->__ptr[i].fileId, 
				     (char *)"Done", &prsDone );
	    if ( cfg->srmDebug ) 
	      cout << " Set request.fileId=" 
		   << prs->requestId <<"."<< prs->fileStatuses->__ptr[i].fileId 
		   << " to Done "<< endl;
	  }
	}
      }
      if ( prsDone )
	prs = prsDone; // point request status to the place it was requested last time
    }
    
  }else
    errors++;

  //
  // Do transfer
  //
  if ( !errors ) {
    string sTurl;
    string sFile;
    class url *dstURL =0;
    class url *srcURL =0;

    switch ( toDo ) {
    case doGET:
      sTurl  = prs->fileStatuses->__ptr[0].TURL;
      srcURL = new class srmURL( sTurl );          // TURL

      sFile  = argf[1];
      dstURL = new class srmURL( sFile );          // args

      break;
    case doPUT:
      sFile  = argf[0];
      srcURL = new class srmURL( sFile );          // args

      sTurl  = prs->fileStatuses->__ptr[0].TURL;
      dstURL = new class srmURL( sTurl );          // TURL
      break;
    default:
      cout << "srmcp(), internal error - toDo switch ... exiting" << endl;
      errors++;
      goto Exit;
    }

    if ( cfg->srmDebug ) {
      cout << "DBG: from=" << srcURL->getURL() << endl << flush;
      cout << "DBG: to=  " << dstURL->getURL() << endl << flush;
    }

    char *exArgs[30];
    int   idx;
    int rc;
    const int sz = 21;
    static char cBfSize[sz]; 
    ostrstream os(cBfSize, sz, ios::out);

    os << cfg->bufferSize << ends; // Convert buffer size to ascii string and terminate by '0'

    string srcHP("");
    string dstHP("");
    //---
    srcHP = srcURL->host();
    if( srcHP == ""
	&& srcURL->scheme() == "file" )
      srcHP = "localhost";

    if( srcURL->port() != "" )
      srcHP += ":" + srcURL->port();
    //---
    dstHP = dstURL->host(); 
    if( dstURL->port() != "" )
      dstHP += ":" + dstURL->port();

    if( dstHP == ""
	&& dstURL->scheme() == "file" )
      dstHP = "localhost";
    //--- [ 0] -- LEN
    idx = 0;

//     cout << "DBG: orig ex=" << exArgs[0] << endl;
//     exArgs[ idx++] = "/bin/echo";

    exArgs[idx++] = (char *)cfg->urlCopyScript.c_str();

    exArgs[idx++] = "-debug";
    exArgs[idx++] = (char *)(( cfg->srmDebug ) ? "true" : "false" );
//     //----------- [  ]
//     exArgs[idx++] = "-x509_user_proxy";
//     exArgs[idx++] = (char *)cfg->x509_userProxy.c_str();

//     exArgs[idx++] = "-x509_user_key";
//     exArgs[idx++] = (char *)cfg->x509_userKey.c_str();

//     exArgs[idx++] = "-x509_user_cert";
//     exArgs[idx++] = (char *)cfg->x509_userCert.c_str();

//     exArgs[idx++] = "-x509_user_certs_dir";
//     exArgs[idx++] = (char *)cfg->x509_certDir.c_str();
    //-----------
    if( cfg->gsiFtpClient == "kftp" ){
      exArgs[idx++] = "-use-kftp";
      exArgs[idx++] = "true";
    }
    //-----------
    exArgs[idx++] = "-buffer_size";
    exArgs[idx++] = (char *)cBfSize;
    //-----------
    exArgs[idx++] = "-src-protocol";
    exArgs[idx++] = (char *)(srcURL->scheme().c_str());

    exArgs[idx++] = "-src-host-port";
    exArgs[idx++] = (char *)srcHP.c_str();

    exArgs[idx++] = "-src-path";
    exArgs[idx++] = (char *)srcURL->urlPath().c_str();
    //-----------
    exArgs[idx++] = "-dst-protocol";
    exArgs[idx++] = (char *)(dstURL->scheme().c_str());

    exArgs[idx++] = "-dst-host-port";
    exArgs[idx++] = (char *)dstHP.c_str();

    exArgs[idx++] = "-dst-path";
    exArgs[idx++] = (char *)dstURL->urlPath().c_str();
    //-----------
    exArgs[idx++] = NULL;

    //===============
    pid_t pid;

    pid = fork ();
    if( pid < 0 ){
      perror("srmcp()");
      errors++;
      goto Exit;
    }
    if ( pid ) { // got child pid, this is a parent
      int stat;
      int rc;

      rc = wait( &stat );
      if( rc < 0) { /* now wait for the child */
	if(errno != ECHILD) 
	  cerr << "srmcp(): An error occurred waiting for children" <<  endl;
      }else if ( rc > 0 ) {
	retCode = stat;	
	if ( cfg->srmDebug ) 
	  cout << "srmcp() Parent: Child process done, status =" << stat << endl << flush;
      }else{
	retCode = 0;
	if ( cfg->srmDebug ) 
	  cout << "srmcp() Parent: Child process done" << endl << flush;
      }

    }else{ //<---- Child process

      if ( cfg->srmDebug )
	cout << "Child: Started ... " << endl << flush;

      rc = execvp ( executable,  exArgs );

      // Will return only IF error
      if ( rc ) { 
	perror( "srmcp()" );
	cerr << "$PATH=" << getenv("PATH") << endl;
	cerr << "srmcp(): can not execvp() transfer process, exiting" << endl;
	exit(1);
      }
    } //<--- End of Child process

    delete srcURL;
    delete dstURL;
  }


 Exit:;
  // Scan Release files in the request if it was not done yet
  if ( prs && string(prs->state) != "Done" ) { 
    for( i=0; i < prs->fileStatuses->__size; i++ ) {
      fileState = prs->fileStatuses->__ptr[i].state;
      if( fileState != "Done" ) {
	if ( cfg->srmDebug ) 
	  cout << " Set request.fileId=" 
	       << prs->requestId <<"."<< prs->fileStatuses->__ptr[i].fileId 
	       << " to Done "<< endl;
	ret = cs->setFileStatus( prs->requestId, prs->fileStatuses->__ptr[i].fileId, 
				 (char *)"Done", &prsDone );
      }
    }
  }

  if ( cfg->srmDebug ) 
    cout << " Done "<< endl;

  delete u1;
  delete u2;
  delete cs;
  delete cfg;

  return (errors) ? 1:retCode;
}
